### PR TITLE
Improve VSOCK communication security by introducing built-in TLS connection blocks

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

Without additional validation, it is not possible for an agent in the guest to find out if a connection attempt comes from `virt-handler`, another privileged service, or from an unprivileged random user. This stems from the fact that while listening on ports `<=1024` requires `CAP_NET_BIND_SERVICE`, connecting to port `<=1024` does not require any privileges.

We pushed for example to `containerd` a change that any containerized workload can not do these types of connections without privileges, but a random logged in user to a node could still connect.
If an agent now provides an interface which allows for instance injecting public ssh keys, any random user could inject their keys into all VMs if they gain access to a node.

To prevent this  type of attack, the following improvments are introduced:

 * A virt-handler `system` grpc service is exposed by virt-handler on port `1`.
 * This type of service requires `CAP_NET_BIND_SERVICE`, so arbitrary users can not provide this service 
 * This `system` services exposes a `CA` method which allows clients to fetch the current kubevirt CA bundle
 * A `tls` parameter is added to the vsock-websocket  all to request a TLS encrypted connection over vsock
 * TLS encryption over virt-handler will be the default. If a custom encryption layer is going to be used the parameter can be set to false.
 * virt-handler will at least right now not attempt a server-certificate validation.

A vsock client can now:

 * expose a service on a port `<=1024`
 * fetch the current CA bundle from the `system` service on every connection attempt
 * and verify the certificate against that CA bundle

Given that both, client and server were as part of this authentication flow listening on a port `<=1024` there is a pretty good guarantee that two services with `CAP_NET_BIND_SERVICE` (one from the node view, the other one from the guest OS view) are talking to each other. This makes it hard for arbitrary users to take such session over.

Finally an example guest agent is included, to demonstrate on how to implement this type of flow.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add basic TLS encryption support for vsock websocket connections
```

### CODE REVIEW

**Comment by rmohr** [comment]
@zhuchenwang FYI, would be great if you could have a look.

File: cmd/example-guest-agent/main.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+			config = &tls.Config{
+				MinVersion: tls.VersionTLS13,
+				ClientCAs:  certPool,
+				GetCertificate: func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+					return cert, nil
+				},
+				// Neither the client nor the server should validate anything itself, `VerifyPeerCertificate` is still executed
+				InsecureSkipVerify: true,
+				// XXX: We need to verify the cert ourselves because we don't have DNS or IP on the certs at the moment
+				VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply cmd/example-guest-agent/main.go:144]: Exactly. Also note that this TLS config is re-created on every connection attempt, which means that `refreshBundle` above is called on every connect, which in turn means that we always have the latest CA bundle. 

Also we know that if virt-handler reaches out to us while we are not able to reaching the system service, something is potentially not right and we just fail without retrying to reach the system service. If virt-handler reaches out, the system service should be available as well.

At the end it is up to vendors what model they prefer, but that allows to continue having a pretty passive agent which can just wait for being contacted.

[the remark this answers is not in the document]
Comment (zhuchenwang) [reply cmd/example-guest-agent/main.go:144]: Does it also work if I have a background goroutine which periodically refreshes the CA bundle from virt-handler?

[the remark this answers is not in the document]
Comment (rmohr) [reply cmd/example-guest-agent/main.go:144]: Hm, you would have to refresh fast to not have a window where you did not yet update the ca. Getting the ca is very cheap compared to the whole websicket flow. I would definitely use it as described here.

File: pkg/virt-handler/vsock/server.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	h.server.Stop()
+	<-h.doneChan
+	h.running = false
+	log.DefaultLogger().Infof("VSOCK server shut down.")
+}
+
+func (h *Hypervisor) start() {
+	log.DefaultLogger().Infof("Starting VSOCK server ...")
+	defer close(h.doneChan)
+	wait.PollInfinite(1*time.Second, func() (done bool, err error) {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-handler/vsock/server.go]: You are right, my implementation was not consistent. What I wanted is retrying no matter if there is an error or not, until the stop channel is closed. Fixed.

File: cmd/example-guest-agent/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhuchenwang) [reply cmd/example-guest-agent/main.go:144]: Make sense. Thanks for the explanation.

**Review by zhuchenwang — COMMENTED** [review]
LGTM

File: cmd/example-guest-agent/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        "//pkg/certificates/triple/cert:go_default_library",
+        "//pkg/vsock/system/v1:go_default_library",
+        "//staging/src/kubevirt.io/client-go/log:go_default_library",
+        "//vendor/github.com/mdlayher/vsock:go_default_library",
+        "@org_golang_google_grpc//:go_default_library",
+    ],
+)
+
+go_binary(
+    name = "example-hook-sidecar",
```
Comment (rthallisey) [inline cmd/example-guest-agent/BUILD.bazel:19]: s/example-hook-sidecar/example-guest-agent/

File: cmd/example-guest-agent/main.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+func main() {
+	log.InitializeLogging("example-guest-agent")
+
+	serverCert, err := setupCert()
+	if err != nil {
+		log.DefaultLogger().Reason(err).Errorf("count not generate the server TLS certificate")
+		os.Exit(1)
+	}
+
+	vsockConn, err := vsock.Listen(1234, &vsock.Config{})
```
Comment (rthallisey) [inline cmd/example-guest-agent/main.go:49]: port 1?

File: staging/src/kubevirt.io/client-go/kubecli/vmi.go [hunk 5]
```diff
@@ -499,5 +499,6 @@ func (v *vmis) VSOCK(name string, options *v1.VSOCKOptions) (StreamInterface, er
 	}
 	queryParams := url.Values{}
 	queryParams.Add("port", strconv.FormatUint(uint64(options.TargetPort), 10))
+	queryParams.Add("tls", strconv.FormatBool(options.UseTLS))
```
Comment (rthallisey) [inline staging/src/kubevirt.io/client-go/kubecli/vmi.go:502]: Can we default to true?

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/kubecli/vmi.go]: Done. It is now the default on the API and in the client. People have to disable it now explicitly. Then they can also add custom e2e encryption.

File: cmd/example-guest-agent/main.go (hunk 4 above)
Comment (rmohr) [reply cmd/example-guest-agent/main.go]: The agent is now used for the guest-agent and the current port settings should be consistent where needed. In general this is the guest-agent port and not the port of the system service (that is probably the confusion here).

**Comment by rmohr** [comment]
@rthallisey happy if you can have another look.

**Comment by rthallisey** [comment]
Change lgtm.  Will add the label after a rebase.

**Comment by rmohr** [comment]
/test  pull-kubevirt-e2e-k8s-1.24-sig-compute

unrelated flake

**Comment by rmohr** [comment]
/test pull-kubevirt-e2e-k8s-1.24-sig-compute-root

failed on an unrelated flake.

**Pull request merged**